### PR TITLE
Add new @case directive that supports lower() and upper() functions

### PR DESCRIPTION
**Pull request by LLuque-twilio**
This PR addresses [JIRA ACT-92](https://segment.atlassian.net/browse/ACT-92)

In it, we add a new `@case` directive that will allow mapping a string from its value in the incoming event to its equivalent `toLowercase()` or `toUppercase()` value. This change will have to be accompanied by a change in `app` that supports these new directives, but that is non-blocking since this PR needs to be implemented first.

App PR `need approval!` https://github.com/segmentio/app/pull/14554

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

Added unit tests to ensure correct functionality.

Did stage deploy of integrations with these changes bundled in & ran a `yarn add @segment/actions-core`. Manually ran a `graphql` request to update the `city` field to use the new `@case` directive. Clearly the UI is not currently equipped to know how to render this new directive, but the event-preview can give you the correct value as shown below. The `city` field is not using the new directive and thus its value is "ATLANTA", whereas `city` is using the new `@case` with `operator:lower` so its value is "atlanta"

Screen Shot 2022-10-25 at 7 09 05 PM

I also did the same to an `Actions Slack` destination and then sent an event through dispatch to see if it was handled correctly. It was!

| Event sent to slack destination  | Resolved message appearing in slack |
| ------------- | ------------- |
| Screen Shot 2022-10-25 at 7 26 47 PM | Screen Shot 2022-10-25 at 7 27 18 PM |

-------

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/core/src/mapping-kit/validate.ts [hunk 1]
```diff
@@ -105,6 +105,16 @@ function validateDirectiveOrString(v: unknown, stack: string[] = []) {
   }
 }
 
+function validateString(v: unknown, stack: string[] = []) {
+  const type = realTypeOrDirective(v)
```
Comment (NolanChan) [inline packages/core/src/mapping-kit/validate.ts:109]: I think if it's just one case, it'd be a bit clearer to do an `if/else`!

File: packages/core/src/mapping-kit/__tests__/index.test.ts [hunk 2]
```diff
@@ -203,6 +203,78 @@ describe('@if', () => {
...
+        '@case': {
+          operator: 'lower',
+          value: { '@path': '$.f' }
+        }
+      },
+      payload
+    )
+    expect(output).toStrictEqual(undefined)
+  })
+})
```
Comment (NolanChan) [inline packages/core/src/mapping-kit/__tests__/index.test.ts:276]: Could we have a test handling an error case? :)

**Review by NolanChan — COMMENTED** [review]
Besides two suggestions, looks good overall! Will wait for a staging test and then will approve!

**Review by dlasky — CHANGES_REQUESTED** [review]

**Comment by dlasky** [comment]
Very solid initial start, I have a few nits around the testing specifically, as well as my slightly more important comment around what to do in the case where we have an unexpected type. In general I think mapping kit should be pretty forgiving so we dont lose data potentially due to a strangely formatted incoming segment event.

File: packages/core/src/mapping-kit/validate.ts (hunk 1 above)
Comment (NolanChan) [reply packages/core/src/mapping-kit/validate.ts:109]: but that should go into `registerDirective` right?

Comment (LLuque-twilio) [reply packages/core/src/mapping-kit/validate.ts:109]: Think you are right here Nolan, the check for the specific operator will go in `registerDirective`, this should just be to ensure that the operator is a string in general!

File: packages/core/src/mapping-kit/__tests__/index.test.ts [hunk 3]
```diff
@@ -203,6 +203,165 @@ describe('@if', () => {
   })
 })
 
+describe.only('@case', () => {
```
Comment (NolanChan) [inline packages/core/src/mapping-kit/__tests__/index.test.ts:206]: left an `.only` :)

File: packages/core/src/mapping-kit/__tests__/index.test.ts [hunk 4]
```diff
@@ -203,42 +203,87 @@ describe('@if', () => {
...
+    let output = transform(
+      {
+        '@case': {
+          operator: 'lower',
+          value: { '@path': '$.a' }
+        }
+      },
+      payload
+    )
+    console.log('number', output)
```
Comment (NolanChan) [inline packages/core/src/mapping-kit/__tests__/index.test.ts:285]: and `console.log`'s

File: packages/core/src/mapping-kit/__tests__/index.test.ts [hunk 5]
```diff
@@ -203,6 +203,162 @@ describe('@if', () => {
...
+          operator: 'upper',
+          value: { '@path': '$.g' }
+        }
+      },
+      payload
+    )
+    expect(output).toStrictEqual('MIXED STRING CASING W/ NUMBERS 190WDB')
+  })
+
+  test('handle non-string-input correctly', () => {
```
Comment (NolanChan) [inline packages/core/src/mapping-kit/__tests__/index.test.ts:275]: minor nit: clarify the test description so we know exactly what should happen, like "handle non-string-input by not modifying the value"

File: packages/core/src/mapping-kit/index.ts [hunk 6]
```diff
@@ -74,6 +74,34 @@ registerDirective('@if', (opts, payload) => {
   }
 })
 
+registerDirective('@case', (opts, payload) => {
+  // let condition = false
```
Comment (NolanChan) [inline packages/core/src/mapping-kit/index.ts:78]: do we need this comment?

**Review by ricardo-rossi — APPROVED** [review]
lgtm

**Pull request merged**